src/native/eventpipe/ep-provider.c

```diff
@@ -427,6 +440,19 @@ provider_invoke_callback (EventPipeProviderCallbackData *provider_callback_data)
 			callback_data /* CallbackContext */);
 	}
 
+	// The callback completed, can take the lock again.
```

Comment:
Couldn't this be done inside ep_provider_callback_data_fini to prevent logic to spread out? Provider callback data will always be allocated/inited and when done used fini/freed after being used.

Reply:
It looks like some places which call fini hold the lock and others do not hold the lock. Does EP_LOCK_ENTER support re-entrancy? I didn't recall that we made use of recursive locking anywhere else.

Reply:
No we should not use recursive locking. init/fini is normally not aware of locking since its setting up the local instance, so maybe it will be problematic to actually introduce that by moving code into fini that needs locking.

Reply:
So then fini is probably not a good place to lock/decrease/signal.

Reply:
My initial thought was to try to keep the increase/decrease closer together and not spread out.

Reply:
Lets save this for future potential refactoring if pattern expands for some reasons.